File: app/src/androidTest/java/com/sdex/activityrunner/LaunchParamsToIntentConverterTest.java
package com.sdex.activityrunner;

import android.content.ComponentName;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.test.filters.MediumTest;
import androidx.test.runner.AndroidJUnit4;

import com.sdex.activityrunner.extensions.IntentExtensionsKt;
import com.sdex.activityrunner.intent.LaunchParams;
import com.sdex.activityrunner.intent.LaunchParamsExtensionsKt;
import com.sdex.activityrunner.intent.LaunchParamsExtra;
import com.sdex.activityrunner.intent.LaunchParamsExtraType;
import com.sdex.activityrunner.intent.converter.LaunchParamsToIntentConverter;
import com.sdex.activityrunner.intent.param.Category;
import com.sdex.activityrunner.intent.param.Flag;

import org.junit.Assert;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

@MediumTest
@RunWith(AndroidJUnit4.class)
public class LaunchParamsToIntentConverterTest {

    @Test
    public void testAction() {
        LaunchParams launchParams = new LaunchParams();
        launchParams.setAction(Intent.ACTION_VIEW);

        LaunchParamsToIntentConverter converter = new LaunchParamsToIntentConverter(launchParams);
        Intent intent = converter.convert();

        Assert.assertEquals(Intent.ACTION_VIEW, intent.getAction());
    }

    @Test
    public void testCategories() {
        LaunchParams launchParams = new LaunchParams();
        launchParams.setCategories(getCategories());

        LaunchParamsToIntentConverter converter = new LaunchParamsToIntentConverter(launchParams);
        Intent intent = converter.convert();

        assertCategoriesEquals(launchParams, intent);
    }

    @Test
    public void testData() {
        LaunchParams launchParams = new LaunchParams();
        launchParams.setData("test_data_string");

        LaunchParamsToIntentConverter converter = new LaunchParamsToIntentConverter(launchParams);
        Intent intent = converter.convert();

        Assert.assertEquals(Uri.parse("test_data_string"), intent.getData());
    }

    @Test
    public void testMimeType() {
        LaunchParams launchParams = new LaunchParams();
        launchParams.setData("data");
        launchParams.setMimeType("image/png");

        LaunchParamsToIntentConverter converter = new LaunchParamsToIntentConverter(launchParams);
        Intent intent = converter.convert();

        Assert.assertEquals("image/png", intent.getType());
    }

    @Test
    public void testClassName() {
        LaunchParams launchParams = new LaunchParams();
        launchParams.setClassName("test_class_name");

        LaunchParamsToIntentConverter converter = new LaunchParamsToIntentConverter(launchParams);
        Intent intent = converter.convert();

        Assert.assertNull(intent.getComponent());
    }

    @Test
    public void testPackageName() {
        LaunchParams launchParams = new LaunchParams();
        launchParams.setPackageName("test_package_name");

        LaunchParamsToIntentConverter converter = new LaunchParamsToIntentConverter(launchParams);
        Intent intent = converter.convert();

        Assert.assertNull(intent.getComponent());
    }

    @Test
    public void testClassAndPackageName() {
        LaunchParams launchParams = new LaunchParams();
        launchParams.setClassName("test_class_name");
        launchParams.setPackageName("test_package_name");

        ComponentName expected = new ComponentName("test_package_name", "test_class_name");

        LaunchParamsToIntentConverter converter = new LaunchParamsToIntentConverter(launchParams);
        Intent intent = converter.convert();

        Assert.assertEquals(expected, intent.getComponent());
    }

    @Test
    public void testExtras() {
        LaunchParams launchParams = new LaunchParams();
        launchParams.setExtras(getExtras());

        LaunchParamsToIntentConverter converter = new LaunchParamsToIntentConverter(launchParams);
        Intent intent = converter.convert();

        assertExtrasEquals(intent);
    }

    @Test
    public void testFlags() {
        LaunchParams launchParams = new LaunchParams();
        launchParams.setFlags(getFlags());

        LaunchParamsToIntentConverter converter = new LaunchParamsToIntentConverter(launchParams);
        Intent intent = converter.convert();

        assertFlagsEquals(launchParams, intent);
    }

    @Test
    public void testSetFrom() {
        LaunchParams launchParams = new LaunchParams();
        launchParams.setAction(Intent.ACTION_ASSIST);
        launchParams.setClassName("cls_name");
        launchParams.setPackageName("pkg_name");
        launchParams.setData("data");
        launchParams.setMimeType("type");
        launchParams.setFlags(getFlags());
        launchParams.setCategories(getCategories());
        launchParams.setExtras(getExtras());

        LaunchParams launchParams2 = new LaunchParams();
        launchParams2.setFrom(launchParams);

        Assert.assertEquals(launchParams, launchParams2);
    }

    @Test
    public void testAll() {
        ComponentName expected = new ComponentName("pkg_name", "cls_name");

        LaunchParams launchParams = new LaunchParams();
        launchParams.setAction(Intent.ACTION_ASSIST);
        launchParams.setClassName(expected.getClassName());
        launchParams.setPackageName(expected.getPackageName());
        launchParams.setData("data");
        launchParams.setMimeType("type");
        launchParams.setFlags(getFlags());
        launchParams.setCategories(getCategories());
        launchParams.setExtras(getExtras());

        LaunchParamsToIntentConverter converter = new LaunchParamsToIntentConverter(launchParams);
        Intent intent = converter.convert();

        Assert.assertEquals(Intent.ACTION_ASSIST, intent.getAction());
        Assert.assertEquals(expected, intent.getComponent());
        Assert.assertEquals(Uri.parse("data"), intent.getData());
        Assert.assertEquals("type", intent.getType());
        assertFlagsEquals(launchParams, intent);
        assertCategoriesEquals(launchParams, intent);
        assertExtrasEquals(intent);
    }

    private static void assertCategoriesEquals(LaunchParams launchParams, Intent intent) {
        Set<String> intentCategories = intent.getCategories();
        List<String> categoriesValues = Category.INSTANCE.list(
                LaunchParamsExtensionsKt.getCategoriesValues(launchParams));

        Assert.assertEquals(categoriesValues.size(), intentCategories.size());

        for (String category : intentCategories) {
            Assert.assertTrue(categoriesValues.contains(category));
        }
    }

    private static void assertFlagsEquals(LaunchParams launchParams, Intent intent) {
        List<String> intentFlags = IntentExtensionsKt.getFlagsList(intent);
        for (Integer position : launchParams.getFlags()) {
            String flagName = Flag.INSTANCE.list().get(position);
            Assert.assertTrue(intentFlags.contains(flagName));
        }
    }

    private static void assertExtrasEquals(Intent intent) {
        Bundle intentExtras = intent.getExtras();

        Assert.assertNotNull(intentExtras);

        Assert.assertEquals("str", intentExtras.getString("k00"));
        Assert.assertEquals("str233", intentExtras.getString("k01"));

        Assert.assertEquals(1, intentExtras.getInt("k10"));
        Assert.assertEquals(0, intentExtras.getInt("k11"));
        Assert.assertEquals(-1, intentExtras.getInt("k12"));

        Assert.assertEquals(1, intentExtras.getLong("k20"));
        Assert.assertEquals(0, intentExtras.getLong("k21"));
        Assert.assertEquals(-1, intentExtras.getLong("k22"));

        Assert.assertEquals(Float.floatToIntBits(0.0f),
                Float.floatToIntBits(intentExtras.getFloat("k30")));
        Assert.assertEquals(Float.floatToIntBits(1.0f),
                Float.floatToIntBits(intentExtras.getFloat("k31")));
        Assert.assertEquals(Float.floatToIntBits(0.1f),
                Float.floatToIntBits(intentExtras.getFloat("k32")));
        Assert.assertEquals(Float.floatToIntBits(-0.2f),
                Float.floatToIntBits(intentExtras.getFloat("k33")));

        Assert.assertEquals(Double.doubleToLongBits(0.0),
                Double.doubleToLongBits(intentExtras.getDouble("k40")));
        Assert.assertEquals(Double.doubleToLongBits(1.0),
                Double.doubleToLongBits(intentExtras.getDouble("k41")));
        Assert.assertEquals(Double.doubleToLongBits(0.1),
                Double.doubleToLongBits(intentExtras.getDouble("k42")));
        Assert.assertEquals(Double.doubleToLongBits(-0.2),
                Double.doubleToLongBits(intentExtras.getDouble("k43")));

        Assert.assertTrue(intentExtras.getBoolean("k50"));
        Assert.assertFalse(intentExtras.getBoolean("k51"));
        Assert.assertFalse(intentExtras.getBoolean("k52"));
        Assert.assertFalse(intentExtras.getBoolean("k53"));
    }

    @NonNull
    private ArrayList<Integer> getCategories() {
        ArrayList<Integer> categories = new ArrayList<>();
        categories.add(Category.INSTANCE.list().indexOf("CATEGORY_APP_BROWSER"));
        categories.add(Category.INSTANCE.list().indexOf("CATEGORY_DEFAULT"));
        categories.add(Category.INSTANCE.list().indexOf("CATEGORY_LAUNCHER"));
        categories.add(Category.INSTANCE.list().indexOf("CATEGORY_PREFERENCE"));
        return categories;
    }

    @NonNull
    private ArrayList<Integer> getFlags() {
        ArrayList<Integer> flags = new ArrayList<>();
        flags.add(Flag.INSTANCE.list().indexOf("FLAG_ACTIVITY_CLEAR_TASK"));
        flags.add(Flag.INSTANCE.list().indexOf("FLAG_ACTIVITY_NEW_TASK"));
        flags.add(Flag.INSTANCE.list().indexOf("FLAG_ACTIVITY_NO_HISTORY"));
        flags.add(Flag.INSTANCE.list().indexOf("FLAG_ACTIVITY_SINGLE_TOP"));
        flags.add(Flag.INSTANCE.list().indexOf("FLAG_GRANT_READ_URI_PERMISSION"));
        return flags;
    }

    @NonNull
    private ArrayList<LaunchParamsExtra> getExtras() {
        ArrayList<LaunchParamsExtra> extras = new ArrayList<>();
        extras.add(new LaunchParamsExtra("k00", "str", LaunchParamsExtraType.STRING, false));
        extras.add(new LaunchParamsExtra("k01", "str233", LaunchParamsExtraType.STRING, false));

        extras.add(new LaunchParamsExtra("k10", "1", LaunchParamsExtraType.INT, false));
        extras.add(new LaunchParamsExtra("k11", "0", LaunchParamsExtraType.INT, false));
        extras.add(new LaunchParamsExtra("k12", "-1", LaunchParamsExtraType.INT, false));

        extras.add(new LaunchParamsExtra("k20", "1", LaunchParamsExtraType.LONG, false));
        extras.add(new LaunchParamsExtra("k21", "0", LaunchParamsExtraType.LONG, false));
        extras.add(new LaunchParamsExtra("k22", "-1", LaunchParamsExtraType.LONG, false));

        extras.add(new LaunchParamsExtra("k30", "0.0", LaunchParamsExtraType.FLOAT, false));
        extras.add(new LaunchParamsExtra("k31", "1.0", LaunchParamsExtraType.FLOAT, false));
        extras.add(new LaunchParamsExtra("k32", "0.1", LaunchParamsExtraType.FLOAT, false));
        extras.add(new LaunchParamsExtra("k33", "-0.2", LaunchParamsExtraType.FLOAT, false));

        extras.add(new LaunchParamsExtra("k40", "0.0", LaunchParamsExtraType.DOUBLE, false));
        extras.add(new LaunchParamsExtra("k41", "1.0", LaunchParamsExtraType.DOUBLE, false));
        extras.add(new LaunchParamsExtra("k42", "0.1", LaunchParamsExtraType.DOUBLE, false));
        extras.add(new LaunchParamsExtra("k43", "-0.2", LaunchParamsExtraType.DOUBLE, false));

        extras.add(new LaunchParamsExtra("k50", "true", LaunchParamsExtraType.BOOLEAN, false));
        extras.add(new LaunchParamsExtra("k51", "false", LaunchParamsExtraType.BOOLEAN, false));
        extras.add(new LaunchParamsExtra("k52", "kek", LaunchParamsExtraType.BOOLEAN, false));
        extras.add(new LaunchParamsExtra("k53", "33", LaunchParamsExtraType.BOOLEAN, false));
        return extras;
    }
}

File: app/src/androidTest/java/com/sdex/activityrunner/LaunchParamsToIntentConverterTest.kt
package com.sdex.activityrunner

import android.content.ComponentName
import android.content.Intent
import android.net.Uri
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.filters.MediumTest
import com.sdex.activityrunner.extensions.getFlagsList
import com.sdex.activityrunner.intent.LaunchParams
import com.sdex.activityrunner.intent.LaunchParamsExtra
import com.sdex.activityrunner.intent.LaunchParamsExtraType
import com.sdex.activityrunner.intent.converter.LaunchParamsToIntentConverter
import com.sdex.activityrunner.intent.getCategoriesValues
import com.sdex.activityrunner.intent.param.Category
import com.sdex.activityrunner.intent.param.Flag
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import org.junit.runner.RunWith

@MediumTest
@RunWith(AndroidJUnit4::class)
class LaunchParamsToIntentConverterTest {

    @Test
    fun convertSetsExplicitAction() {
        val intent = LaunchParams(action = Intent.ACTION_VIEW).toIntent()

        assertEquals(Intent.ACTION_VIEW, intent.action)
    }

    @Test
    fun convertUsesMainActionWhenActionIsNotSet() {
        val intent = LaunchParams().toIntent()

        assertEquals(Intent.ACTION_MAIN, intent.action)
    }

    @Test
    fun convertUsesMainActionWhenActionIsEmpty() {
        val intent = LaunchParams(action = "").toIntent()

        assertEquals(Intent.ACTION_MAIN, intent.action)
    }

    @Test
    fun convertAddsCategories() {
        val launchParams = LaunchParams(categories = categories)
        val intent = launchParams.toIntent()

        assertCategoriesEquals(launchParams, intent)
    }

    @Test
    fun convertDoesNotAddCategoriesWhenCategoriesAreEmpty() {
        val intent = LaunchParams(categories = emptyList()).toIntent()

        assertNull(intent.categories)
    }

    @Test
    fun convertSetsData() {
        val intent = LaunchParams(data = "test_data_string").toIntent()

        assertEquals(Uri.parse("test_data_string"), intent.data)
    }

    @Test
    fun convertSetsMimeTypeWhenDataIsSet() {
        val intent = LaunchParams(
            data = "data",
            mimeType = "image/png",
        ).toIntent()

        assertEquals("image/png", intent.type)
    }

    @Test
    fun convertIgnoresMimeTypeWhenDataIsNotSet() {
        val intent = LaunchParams(mimeType = "image/png").toIntent()

        assertNull(intent.data)
        assertNull(intent.type)
    }

    @Test
    fun convertIgnoresDataAndMimeTypeWhenDataIsEmpty() {
        val intent = LaunchParams(
            data = "",
            mimeType = "image/png",
        ).toIntent()

        assertNull(intent.data)
        assertNull(intent.type)
    }

    @Test
    fun convertSetsPackageWithoutComponentWhenOnlyPackageNameIsSet() {
        val intent = LaunchParams(packageName = "test_package_name").toIntent()

        assertEquals("test_package_name", intent.`package`)
        assertNull(intent.component)
    }

    @Test
    fun convertIgnoresClassNameWhenPackageNameIsMissing() {
        val intent = LaunchParams(className = "test_class_name").toIntent()

        assertNull(intent.`package`)
        assertNull(intent.component)
    }

    @Test
    fun convertIgnoresEmptyPackageName() {
        val intent = LaunchParams(packageName = "").toIntent()

        assertNull(intent.`package`)
        assertNull(intent.component)
    }

    @Test
    fun convertDoesNotSetComponentWhenClassNameIsEmpty() {
        val intent = LaunchParams(
            packageName = "test_package_name",
            className = "",
        ).toIntent()

        assertEquals("test_package_name", intent.`package`)
        assertNull(intent.component)
    }

    @Test
    fun convertSetsComponentWhenClassAndPackageNamesAreSet() {
        val expected = ComponentName("test_package_name", "test_class_name")
        val intent = LaunchParams(
            packageName = expected.packageName,
            className = expected.className,
        ).toIntent()

        assertEquals(expected, intent.component)
    }

    @Test
    fun convertAddsExtras() {
        val intent = LaunchParams(extras = extras).toIntent()

        assertExtrasEquals(intent)
    }

    @Test
    fun convertDoesNotAddExtrasWhenExtrasAreEmpty() {
        val intent = LaunchParams(extras = emptyList()).toIntent()

        assertNull(intent.extras)
    }

    @Test
    fun convertSkipsExtrasWithInvalidNumericValues() {
        val intent = LaunchParams(
            extras = listOf(
                LaunchParamsExtra("int", "not_int", LaunchParamsExtraType.INT, false),
                LaunchParamsExtra("long", "not_long", LaunchParamsExtraType.LONG, false),
                LaunchParamsExtra("float", "not_float", LaunchParamsExtraType.FLOAT, false),
                LaunchParamsExtra("double", "not_double", LaunchParamsExtraType.DOUBLE, false),
            ),
        ).toIntent()

        assertNull(intent.extras)
    }

    @Test
    fun convertAddsFlags() {
        val launchParams = LaunchParams(flags = flags)
        val intent = launchParams.toIntent()

        assertFlagsEquals(launchParams, intent)
    }

    @Test
    fun convertDoesNotAddFlagsWhenFlagsAreEmpty() {
        val intent = LaunchParams(flags = emptyList()).toIntent()

        assertEquals(0, intent.flags)
    }

    @Test
    fun convertMapsAllSupportedFields() {
        val expected = ComponentName("pkg_name", "cls_name")
        val launchParams = LaunchParams(
            action = Intent.ACTION_ASSIST,
            packageName = expected.packageName,
            className = expected.className,
            data = "data",
            mimeType = "type",
            flags = flags,
            categories = categories,
            extras = extras,
        )

        val intent = launchParams.toIntent()

        assertEquals(Intent.ACTION_ASSIST, intent.action)
        assertEquals(expected, intent.component)
        assertEquals(Uri.parse("data"), intent.data)
        assertEquals("type", intent.type)
        assertFlagsEquals(launchParams, intent)
        assertCategoriesEquals(launchParams, intent)
        assertExtrasEquals(intent)
    }

    private fun LaunchParams.toIntent(): Intent = LaunchParamsToIntentConverter(this).convert()

    private fun assertCategoriesEquals(launchParams: LaunchParams, intent: Intent) {
        val intentCategories = requireNotNull(intent.categories)
        val categoriesValues = Category.list(launchParams.getCategoriesValues())

        assertEquals(categoriesValues.size, intentCategories.size)
        assertTrue(intentCategories.all { it in categoriesValues })
    }

    private fun assertFlagsEquals(launchParams: LaunchParams, intent: Intent) {
        val intentFlags = intent.getFlagsList()
        launchParams.flags.forEach { position ->
            val flagName = Flag.list()[position]
            assertTrue(intentFlags.contains(flagName))
        }
    }

    private fun assertExtrasEquals(intent: Intent) {
        val intentExtras = intent.extras
        assertNotNull(intentExtras)

        requireNotNull(intentExtras)
        assertEquals("str", intentExtras.getString("k00"))
        assertEquals("str233", intentExtras.getString("k01"))

        assertEquals(1, intentExtras.getInt("k10"))
        assertEquals(0, intentExtras.getInt("k11"))
        assertEquals(-1, intentExtras.getInt("k12"))

        assertEquals(1L, intentExtras.getLong("k20"))
        assertEquals(0L, intentExtras.getLong("k21"))
        assertEquals(-1L, intentExtras.getLong("k22"))

        assertEquals(0.0f.toBits(), intentExtras.getFloat("k30").toBits())
        assertEquals(1.0f.toBits(), intentExtras.getFloat("k31").toBits())
        assertEquals(0.1f.toBits(), intentExtras.getFloat("k32").toBits())
        assertEquals((-0.2f).toBits(), intentExtras.getFloat("k33").toBits())

        assertEquals(0.0.toBits(), intentExtras.getDouble("k40").toBits())
        assertEquals(1.0.toBits(), intentExtras.getDouble("k41").toBits())
        assertEquals(0.1.toBits(), intentExtras.getDouble("k42").toBits())
        assertEquals((-0.2).toBits(), intentExtras.getDouble("k43").toBits())

        assertTrue(intentExtras.getBoolean("k50"))
        assertFalse(intentExtras.getBoolean("k51"))
        assertFalse(intentExtras.containsKey("k52"))
        assertFalse(intentExtras.containsKey("k53"))
    }

    private val categories: List<Int>
        get() = listOf(
            Category.list().indexOf("CATEGORY_APP_BROWSER"),
            Category.list().indexOf("CATEGORY_DEFAULT"),
            Category.list().indexOf("CATEGORY_LAUNCHER"),
            Category.list().indexOf("CATEGORY_PREFERENCE"),
        )

    private val flags: List<Int>
        get() = listOf(
            Flag.list().indexOf("FLAG_ACTIVITY_CLEAR_TASK"),
            Flag.list().indexOf("FLAG_ACTIVITY_NEW_TASK"),
            Flag.list().indexOf("FLAG_ACTIVITY_NO_HISTORY"),
            Flag.list().indexOf("FLAG_ACTIVITY_SINGLE_TOP"),
            Flag.list().indexOf("FLAG_GRANT_READ_URI_PERMISSION"),
        )

    private val extras: List<LaunchParamsExtra>
        get() = listOf(
            LaunchParamsExtra("k00", "str", LaunchParamsExtraType.STRING, false),
            LaunchParamsExtra("k01", "str233", LaunchParamsExtraType.STRING, false),

            LaunchParamsExtra("k10", "1", LaunchParamsExtraType.INT, false),
            LaunchParamsExtra("k11", "0", LaunchParamsExtraType.INT, false),
            LaunchParamsExtra("k12", "-1", LaunchParamsExtraType.INT, false),

            LaunchParamsExtra("k20", "1", LaunchParamsExtraType.LONG, false),
            LaunchParamsExtra("k21", "0", LaunchParamsExtraType.LONG, false),
            LaunchParamsExtra("k22", "-1", LaunchParamsExtraType.LONG, false),

            LaunchParamsExtra("k30", "0.0", LaunchParamsExtraType.FLOAT, false),
            LaunchParamsExtra("k31", "1.0", LaunchParamsExtraType.FLOAT, false),
            LaunchParamsExtra("k32", "0.1", LaunchParamsExtraType.FLOAT, false),
            LaunchParamsExtra("k33", "-0.2", LaunchParamsExtraType.FLOAT, false),

            LaunchParamsExtra("k40", "0.0", LaunchParamsExtraType.DOUBLE, false),
            LaunchParamsExtra("k41", "1.0", LaunchParamsExtraType.DOUBLE, false),
            LaunchParamsExtra("k42", "0.1", LaunchParamsExtraType.DOUBLE, false),
            LaunchParamsExtra("k43", "-0.2", LaunchParamsExtraType.DOUBLE, false),

            LaunchParamsExtra("k50", "true", LaunchParamsExtraType.BOOLEAN, false),
            LaunchParamsExtra("k51", "false", LaunchParamsExtraType.BOOLEAN, false),
            LaunchParamsExtra("k52", "kek", LaunchParamsExtraType.BOOLEAN, false),
            LaunchParamsExtra("k53", "33", LaunchParamsExtraType.BOOLEAN, false),
        )
}
